### PR TITLE
Better sprite menu

### PR DESCRIPTION
**Pull request by fath11**
Better sprite menu layout. More layouts will be added in the future.

## Scratch's Grid layout:
image

### Pros:
1. Able to fit 20 sprites at once without scrolling.
2. Compact while also being aesthetically pleasing.

### Issues: 
1. Due to how Gandi IDE organizes their default sprite menu layout, its almost impossible for me to achieve vertical stage selector like Scratch have. I did try, but it causes Gandi to crash when it attempts to collapse the sprite menu.
3. For some reason, `text-overflow: ellipsis` style isn't working on the sprite selector's label?

## Compact layout: 
image

### Pros:
1. Able to fit 13 sprites at once without scrolling.
2. Still looks like Gandi IDE's default layout while being able to show A LOT more sprites at once.

### Cons:
1. Text is very small and might be hard or impossible to read for some people.

## Super Compact Layout:
image

### Pros:
1. Able to fit up to **65 sprites** at once without scrolling.

### Cons:
1. Very ugly.
2. Your screen size will impact how many sprites you can show. I can't bother fixing this.
3. Text is very small AND crammed that it might be hard or impossible to read for some people.

### CODE REVIEW

**Comment by fath11** [comment]
I might do a few more polishing and open this PR for review. After that, some people suggest adding custom layout feature to this plugin but i am not sure if that will any useful.

**Comment by fath11** [comment]
image
Ready for review

**Comment by fath11** [comment]
In the future: I am planning to rename this plugin to More Layouts or something else. It will have more layouts for more menus.

**Comment by zxq142857** [comment]
You need to add a Chinese translation.

**Comment by zxq142857** [comment]
In team collaboration mode, the display of collaborators' avatars isn't compatible.

File: src/plugins/better-sprite-menu/index.tsx [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+              options: [
+                { label: msg('plugins.betterSpriteMenu.layouts.default'), value: menuLayoutList[0] },
+                { label: msg('plugins.betterSpriteMenu.layouts.grid'), value: menuLayoutList[1] },
+                { label: msg('plugins.betterSpriteMenu.layouts.compact'), value: menuLayoutList[2] },
+                { label: msg('plugins.betterSpriteMenu.layouts.superCompact'), value: menuLayoutList[3] },
+              ],
+              onChange: (value) => {
+                currentSpriteMenuLayout = value.toString();
+                updateSpriteMenuStyle();
+                console.log(redux)
```
Comment (zxq142857) [inline src/plugins/better-sprite-menu/index.tsx:236]: The debugging code needs to be removed.

**Comment by zxq142857** [comment]
Now, after removing this plugin, the styles won't revert to their original state, so should we remove this side effect upon uninstallation?

**Review by zxq142857 — COMMENTED** [review]
There are some minor code style issues; I'll fix them all with a one-click before the final submission.

**Review by zxq142857 — COMMENTED** [review]
Some areas still need to be refined.

**Comment by fath11** [comment]
> In team collaboration mode, the display of collaborators' avatars isn't compatible. 

my plugin's build file isn't loading in Cocrea, and I don't have a CCW account to use Collaborative mode

**Comment by zxq142857** [comment]
> > In team collaboration mode, the display of collaborators' avatars isn't compatible.
> 
> my plugin's build file isn't loading in Cocrea, and I don't have a CCW account to use Collaborative mode

In the playground of this project, collaboration testing is not possible. You can only package your plugin using `pnpm buildSinglePlugin --name=better-sprite-menu`, then import and use it in the Custom Plugin.
image

**Comment by fath11** [comment]
zh-cn translation need some reviewing

**Review by zxq142857 — COMMENTED** [review]
Looking forward to its launch.

**Comment by fath11** [comment]
zh-cn is updated, need review

File: src/l10n/zh-cn.json [hunk 2]
```diff
@@ -151,5 +151,14 @@
   "plugins.fastInput.description": "借鉴自 ScratchAddons 的 InsertBlocksByName，增加了中文语言下可使用拼音缩写输入特性。\n💡使用方法：在代码区域中按住 Shift 键并点击打开积木输入窗口，输入积木名称或拼音缩写，使用方向键和回车键进行导航，使用 Tab 键进行自动补全，按住 Shift 键拖动积木可防止列表关闭。",
   "plugins.fastInput.popup_scale": "积木缩放",
   "plugins.fastInput.popup_width": "弹窗宽度",
-  "plugins.fastInput.popup_max_height": "弹窗最大高度"
-}
\ No newline at end of file
+  "plugins.fastInput.popup_max_height": "弹窗最大高度",
+  "plugins.betterSpriteMenu.title": "更好的雪碧菜单",
+  "plugins.betterSpriteMenu.title": "Better Character List Layout",
```
Comment (zxq142857) [inline src/l10n/zh-cn.json:156]: This line is repeated and needs to be deleted

File: src/l10n/zh-cn.json [hunk 3]
```diff
@@ -151,5 +151,14 @@
   "plugins.fastInput.description": "借鉴自 ScratchAddons 的 InsertBlocksByName，增加了中文语言下可使用拼音缩写输入特性。\n💡使用方法：在代码区域中按住 Shift 键并点击打开积木输入窗口，输入积木名称或拼音缩写，使用方向键和回车键进行导航，使用 Tab 键进行自动补全，按住 Shift 键拖动积木可防止列表关闭。",
   "plugins.fastInput.popup_scale": "积木缩放",
   "plugins.fastInput.popup_width": "弹窗宽度",
-  "plugins.fastInput.popup_max_height": "弹窗最大高度"
-}
\ No newline at end of file
+  "plugins.fastInput.popup_max_height": "弹窗最大高度",
+  "plugins.betterSpriteMenu.title": "更好的雪碧菜单",
```
Comment (zxq142857) [inline src/l10n/zh-cn.json:155]: It is recommended to translate: "更好的角色列表"

File: src/l10n/zh-cn.json [hunk 4]
```diff
@@ -151,5 +151,14 @@
...
+  "plugins.betterSpriteMenu.title": "Better Character List Layout",
+  "plugins.betterSpriteMenu.layouts": "布局",
+  "plugins.betterSpriteMenu.layouts.label": "布局类型：",
+  "plugins.betterSpriteMenu.layouts.default": "默认",
+  "plugins.betterSpriteMenu.layouts.grid": "原版的Scratch布局",
+  "plugins.betterSpriteMenu.layouts.compact": "紧凑",
+  "plugins.betterSpriteMenu.layouts.superCompact": "超级紧凑",
+  "plugins.betterSpriteMenu.description": "给精灵选择器一个新的（更好的）外观!"
```
Comment (zxq142857) [inline src/l10n/zh-cn.json:163]: It is recommended to translate: "给角色列表一个新的（更好的）外观!"，

Comment (zxq142857) [inline src/l10n/zh-cn.json:163]: In the Chinese version of Scratch, 'Sprite' is translated as '角色' instead of '精灵'.

**Review by zxq142857 — COMMENTED** [review]
There are still some minor issues with the Chinese translation.

**Review by zxq142857 — COMMENTED** [review]
Better sprite menu.

**Pull request merged**